pkg/pushsync/pushsync_test.go

```diff
@@ -8,223 +8,205 @@ import (
 
-	recorder.SetProtocols(ps.Protocol())
+	pivotRecorder := streamtest.New(
+		streamtest.WithProtocols(psPivot.Protocol()),
+		streamtest.WithMiddlewares(func(f p2p.HandlerFunc) p2p.HandlerFunc {
+			return f
+		}),
+	)
 
-	stream, err := recorder.NewStream(context.Background(), pivotNode, nil, pushsync.ProtocolName, pushsync.ProtocolVersion, pushsync.StreamName)
-	if err != nil {
-		t.Fatal(err)
-	}
-	defer stream.Close()
-	w := protobuf.NewWriter(stream)
+	// Creating the trigger peer
+	psTriggerPeer, triggerStorerDB := createPushSyncNode(t, triggerPeer, pivotRecorder, mock.WithClosestPeer(pivotPeer))
 
-	// this triggers the handler of the pivot with a delivery stream
-	err = w.WriteMsg(&pb.Delivery{
-		Address: chunkAddress.Bytes(),
-		Data:    chunkData,
-	})
+	// upload the chunk to the trigger node DB which triggers the chain reaction of sending this chunk
+	// from trigger peer to pivot peer to closest peer.
+	_, err := triggerStorerDB.Put(context.Background(), storage.ModePutUpload, chunk)
 	if err != nil {
 		t.Fatal(err)
 	}
 
-	_ = recorder.WaitRecords(t, closestPeer, pushsync.ProtocolName, pushsync.ProtocolVersion, pushsync.StreamName, 1, 5)
-	mtx.Lock()
-	defer mtx.Unlock()
-	if !targetCalled {
-		t.Fatal("target not called")
-	}
-}
+	// In pivot peer,  intercept the incoming delivery chunk from the trigger peer and check for correctness
+	waitOnRecordAndTest(t, pivotPeer, pivotRecorder, chunkAddress, chunkData)
 
-// TestNoForwardChunk tests that the closest node to a chunk doesn't forward it to other nodes.
-func TestNoForwardChunk(t *testing.T) {
-	logger := logging.New(ioutil.Discard, 0)
+	// Pivot peer will forward the chunk to its closest peer. Intercept the incoming stream from pivot node and check
+	// for the correctness of the chunk
+	waitOnRecordAndTest(t, closestPeer, closestRecorder, chunkAddress, chunkData)
 
-	// chunk data to upload
-	chunkAddress := swarm.MustParseHexAddress("7000000000000000000000000000000000000000000000000000000000000000") // binary 0111
-	chunkData := []byte("1234")
+	// Similarly intercept the same incoming stream to see if the closest peer is sending a proper receipt
+	waitOnRecordAndTest(t, closestPeer, closestRecorder, chunkAddress, nil)
 
-	// create a pivot node and a cluster of nodes
-	pivotNode := swarm.MustParseHexAddress("6000000000000000000000000000000000000000000000000000000000000000") // pivot is 0110
+	// In the received stream, check if a receipt is sent from pivot peer and check for its correctness.
+	waitOnRecordAndTest(t, pivotPeer, pivotRecorder, chunkAddress, nil)
 
-	// Create a mock connectivity
-	mockTopology := mock.NewTopologyDriver(mock.WithClosestPeerErr(topology.ErrWantSelf))
+	// closing the ps so that the DB is closed after that to avoid data race
+	psClosestPeer.Close()
+	psPivot.Close()
+	psTriggerPeer.Close()
+
+	// closing all DBs
+	closestStorerPeerDB.Close()
+	storerPivotDB.Close()
+	triggerStorerDB.Close()
```

Comment:
I think that these storer Close calls can be deferred to make sure that they are called. They will certainly be called after pushsync close calls.